### ISSUE TITLE
issue with negative padding

### ISSUE THREAD

**Issue by SuggonM** [pavle-goloskokovic/texture-unpacker#5]
Hello, good day. I recently ran into this issue where Sharp was complaining about encountering values that are negative.
```
Error: Expected positive integer for bottom but received -9 of type number
```

The json to be read looks like this:
```json
"coffee.png":
{
	"frame": {"x":453,"y":184,"w":48,"h":45},
	"rotated": false,
	"trimmed": false,
	"spriteSourceSize": {"x":8,"y":9,"w":48,"h":45},
	"sourceSize": {"w":48,"h":45}
},
```

I found that based on the calculations used in [unpacker.ts#L311-L312](https://github.com/pavle-goloskokovic/texture-unpacker/blob/master/unpacker.ts#L311-L312), the padding values would actually evaluate to negatives, which Sharp thinks is invalid.
```
    right: -8,
    bottom: -9,
```
---
Although changing `sss.w - sss.x` into `sss.w + sss.x` seemed to locally solve the issue in my case, I'm skeptical about the validity of this change. Could you please re-examine the formula to find if it's correct?

**Comment by pavle-goloskokovic** [comment]
Hi @SuggonM, thanks for reporting this.

I must assume that in this case it is an issue with provided data and not the formula.

`sourceSize` describes the size of the original untrimmed image.

`spriteSourceSize` describes the offset of the actual sprite graphics within the untrimmed image (`x` for left padding and `y` for top padding) and its size.

`trimmed` boolean indicates if original image was trimmed, obviously :)

`frame` describes the position of the sprite within the atlas and its size.

If `trimmed` is `false`, like in your example, then `frame`, `sourceSize`, and `spriteSourceSize` should all have the same `w` and `h` values and `spriteSourceSize` `x` and `y` should be `0`.

If `trimmed` is `true`, then `frame` and `spriteSourceSize` should have the same `w` and `h` values, and `sourceSize`should be at least the sum of padding and size described in `spriteSourceSize`, or greater if there is also right and bottom padding.

You can have a look at the example data to see all possible correct combinations for `rotated` and `trimmed` values here:
https://github.com/pavle-goloskokovic/texture-unpacker/blob/master/example/Sprite.json

My best guess is that someone manually changed `spriteSourceSize` padding values from `0` to current, leading to having inconsistent data and experiencing the issue with Sharp.

Let me know if you have any other ideas or suggestions.

**Comment by SuggonM** [comment]
Hi and thank you for quick response in such great detail.

I can ensure the data hasn't been tampered with as the json itself, to be honest, was extracted raw from the game assets, which I have no control over. The game has been using the same atlas for ages to render all textures so a mapping error doesn't quite justify it either (although the term "for ages" does open a possibility of version incompatibility in-between as TexturePacker is apparently in [v7.1.0](https://www.codeandweb.com/texturepacker) today?).

To add a bit more to the context, I've uploaded the json and texture here just so that it could be investigated further: [worldmap_common.json](https://github.com/pavle-goloskokovic/texture-unpacker/files/14319168/worldmap_common.json) and [worldmap_common.png](https://github.com/pavle-goloskokovic/texture-unpacker/assets/126052298/63ea0546-616d-43e7-be3b-f1b6c213cc3f)

**Comment by pavle-goloskokovic** [comment]
Thanks for providing json data and atlas texture!

I came across this forum post related to TexturePacker trim mode:
https://forum.playcanvas.com/t/texture-packer-trimmed-sprites-and-positioning/15024/3

Seems that logic I described above is valid only if atlas was exported with Trim mode set to `Trim` or `None`.

If Trim mode is set to any of the `Crop` options, then `sourceSize` does not have the value of the original uncropped image, but rather of the cropped sprite.

This means that with the available frame data there is no way to determine original image size unfortunately.
`spriteSourceSize` can still have values for top and left padding so I kept those in to have the output as close to the original as possible, but in case if `trimmed` is set to `false`, right and bottom padding values will always be `0`.

I’ve tested updated logic with the files you provided, and it works well now, as well as with the original example.

I will mark this issue closed as it should be resolved with https://github.com/pavle-goloskokovic/texture-unpacker/commit/6e5c3e36c55b3906e20f2d8850ed7b4734533078.

Feel free to write if there is anything I’ve missed.

Thanks for reporting the issue and helping make this tool better!
